dsgrid/dataset/pivoted_table.py

```diff
@@ -158,39 +171,62 @@ def process_stacked_aggregations(
         if not aggregations:
             return df
 
-        group_by_query_names = []
         pivoted_dimension_type = context.get_pivoted_dimension_type(dataset_id=self.dataset_id)
         pivoted_columns = set(context.get_pivoted_columns(dataset_id=self.dataset_id))
+        final_query_names = {x: set() for x in DimensionType if x != pivoted_dimension_type}
+        column_to_dim_type = {}
+        dropped_dimensions = set()
         for agg in aggregations:
             columns = []
             for dim_type, column in agg.iter_dimensions_to_keep():
+                assert dim_type not in dropped_dimensions, dim_type
                 if dim_type != pivoted_dimension_type:
                     columns.append(column)
+                    self._add_column_to_dim_type(column, dim_type, column_to_dim_type)
+            dropped_dimensions.update(set(agg.list_dropped_dimensions()))
             if not columns:
                 continue
+
             df = self.add_columns(df, columns, context, True)
             group_by_cols = []
             for column in columns:
-                if column.function is None:
-                    expr = column.dimension_query_name
-                else:
-                    expr = column.function(column.dimension_query_name)
-                    if column.alias is not None:
-                        expr = expr.alias(column.alias)
+                label = column.get_column_label()
+                dim_type = column_to_dim_type[label]
+                expr = self._make_group_by_column_expr(column)
+                if not isinstance(expr, str) or expr != column.dimension_query_name:
+                    # In this case we are replacing any existing query name with an expression
+                    # or alias, and so the old name must be removed.
+                    final_query_names[dim_type].discard(column.dimension_query_name)
+                final_query_names[dim_type].add(label)
                 group_by_cols.append(expr)
-                group_by_query_names.append(column.dimension_query_name)
+
             op = agg.aggregation_function
             agg_expr = [op(x).alias(x) for x in df.columns if x in pivoted_columns]
             df = df.groupBy(*group_by_cols).agg(*agg_expr)
             logger.info(
                 "Aggregated dimensions with groupBy %s and agg %s", group_by_cols, agg_expr
             )
 
-        final_query_names = {x: set() for x in DimensionType if x != pivoted_dimension_type}
-        for column in group_by_query_names:
-            dim = self.project_config.get_dimension(column)
-            final_query_names[dim.model.dimension_type].add(column)
-
         for dim_type, columns in final_query_names.items():
+            if dim_type in dropped_dimensions and columns:
```

Comment:
I did some refactoring while making a fix to this function. The main thing is that when the function exits the columns of the dataframe are stored as "dimension query names" in the query context. But they aren't always actual dimension query names.